iavl/iavl_proof2.go

```diff
@@ -0,0 +1,268 @@
+		}
+	}
+
+	return nil
+}
+
+type KeyRangeExistsProof struct {
+	RootHash   []byte
+	PathToKeys []*PathToKey
+}
+
+func (proof *KeyRangeExistsProof) Verify(key []byte, value []byte, root []byte) bool {
+	return false
+}
+
+func ReadKeyExistsProof(data []byte) (*KeyExistsProof, error) {
+	proof := new(KeyExistsProof)
+	err := wire.ReadBinaryBytes(data, &proof)
+	if err != nil {
+		return nil, err
+	}
+	return proof, nil
+}
+
+func (node *IAVLNode) constructKeyExistsProof(t *IAVLTree, key []byte, proof *KeyExistsProof) (value []byte, exists bool) {
+	if node.height == 0 {
+		if bytes.Compare(node.key, key) == 0 {
+			proof.LeafHash = node.hash
+			return node.value, true
+		}
+		return nil, false
+	}
+
+	if bytes.Compare(key, node.key) < 0 {
+		if value, exists := node.getLeftNode(t).constructKeyExistsProof(t, key, proof); exists {
+			branch := IAVLProofInnerNode{
+				Height: node.height,
+				Size:   node.size,
+				Left:   nil,
+				Right:  node.getRightNode(t).hash,
+			}
+			proof.InnerNodes = append(proof.InnerNodes, branch)
+			return value, true
+		}
+		return nil, false
+	}
+
+	if value, exists = node.getRightNode(t).constructKeyExistsProof(t, key, proof); exists {
+		branch := IAVLProofInnerNode{
+			Height: node.height,
+			Size:   node.size,
+			Left:   node.getLeftNode(t).hash,
+			Right:  nil,
+		}
+		proof.InnerNodes = append(proof.InnerNodes, branch)
+		return value, true
+	}
+	return nil, false
+}
+
+func (node *IAVLNode) constructKeyNotExistsProof(t *IAVLTree, key []byte, proof *KeyNotExistsProof) error {
+	idx, _, exists := t.Get(key)
+	if exists {
+		return errors.Errorf("couldn't construct non-existence proof: key 0x%x exists", key)
+	}
+
+	var (
+		lkey, lval []byte
+		rkey, rval []byte
+	)
+	if idx > 0 {
+		lkey, lval = t.GetByIndex(idx - 1)
+	}
+	if idx <= t.Size()-1 {
+		rkey, rval = t.GetByIndex(idx)
+	}
+
+	if lkey == nil && rkey == nil {
+		return errors.New("couldn't get keys required for non-existence proof")
+	}
+
+	if lkey != nil {
+		lproof := &KeyExistsProof{
+			RootHash: t.root.hash,
+		}
+		node.constructKeyExistsProof(t, lkey, lproof)
+		proof.LeftPath = &lproof.PathToKey
+		proof.LeftNode = IAVLProofLeafNode{KeyBytes: lkey, ValueBytes: lval}
+	}
+	if rkey != nil {
+		rproof := &KeyExistsProof{
+			RootHash: t.root.hash,
+		}
+		node.constructKeyExistsProof(t, rkey, rproof)
+		proof.RightPath = &rproof.PathToKey
+		proof.RightNode = IAVLProofLeafNode{KeyBytes: rkey, ValueBytes: rval}
+	}
+
+	return nil
+}
+
+func (t *IAVLTree) ConstructKeyExistsProof(key []byte) (value []byte, proof *KeyExistsProof) {
+	if t.root == nil {
+		return nil, nil
+	}
+	t.root.hashWithCount(t) // Ensure that all hashes are calculated.
+	proof = &KeyExistsProof{
+		RootHash: t.root.hash,
+	}
+	if value, exists := t.root.constructKeyExistsProof(t, key, proof); exists {
+		return value, proof
+	}
+	return nil, nil
+}
+
+func (t *IAVLTree) ConstructKeyNotExistsProof(key []byte) (*KeyNotExistsProof, error) {
```

Comment:
Also vote for `constructKeyNotExistsProof` unless we really need to expose directly.